README.rst
===============================
Cloud Backup Agent Auto-Updater
===============================

Overview
========

Rackspace Cloud Backup Agent resides on a variety of different Linux
distributions, and may expand its support to FreeBSD in a near future.  Unlike
Windows (despite of Windows Store), these platforms all come with some form of
"package management" systems, so that user can upgrade the backup agents
through these systems.  However, such upgrade tasks involve user interactions,
while the practice we provide for Windows involves no user interactions, and
has been proven to be more user-friendly and still robust.  To delivery a more
consistent and comprehensive user-experience package, we launch the auto-
updater support on the Linux distributions.

In addition, due to the difference between the repository support of the
package management systems, the steps to install the agent from scratch on a
Linux box is even more complicated.  This is also a problem which the auto-
updater aims to solve.


Supported Platforms
===================

- Distributions

  1. Redhat flavors

     - Fedora 18, 19, 20
     - RHEL 5, 6
     - CentOS 5, 6
     - Scientific Linux 6

  2. Debian flavors

     - Debian 6, 7
     - Ubuntu 10, 12, 13

- Python versions

  The auto-updater runs on the default Python implementation on each
  distribution, and such an implementation is required to be py2x.  The
  following Python versions are supported:

  2.4, 2.5 [1]_, 2.6, 2.7

.. [1] However, no supported distribution comes with a default Python version
   of 2.5.


Installation
============

- Redhat flavors

  1. Download the latest rpm package;
  2. ``sudo yum install cloudbackup-updater-<version>-2.noarch.rpm``

- Debian flavors

  1. Download the latest deb package;
  2. ``sudo dpkg -i cloudbackup-updater_<version>-2_all.deb``


Behaviors
=========

This section describes the expected behavior of the auto-updater after the
installation under different scenarios.  Scenarios are the valid combinations
of all the conditions.  The exact behaviors are different with different
package management systems, but the logic are the same.

Conditions:

1. The agent repository is correctly configured.

2. An agent is installed.

3. The installed agent is updated.  Requires 2).

4. The installed agent is running.  Requires 2).

5. The running agent is performing a backup, a restore, or a cleanup.
   Requires 4).

Behaviors:

If 1) is not true, then agent repository is configured.

Debian: `/etc/apt/sources.list.d/driveclient.list` is created and the command line output of ``sudo apt-key list`` contains the agent repo key.

Redhat: `/etc/yum.repos.d/drivesrvr.repo` is created.

*[Note:* A better approach to verify the existence of the repositories is
to remove the installed agent then install it:

Debian: ``sudo apt-get remove driveclient``, ``sudo apt-get install driveclient``

Redhat: ``sudo yum remove driveclient``, ``sudo yum install driveclient``

Since there is no local package involved, the installation sources must be the
repositories.

*-- end note]*

If 2) is not true, then the newest agent is installed from the repository.

*[Note:* The installation or upgrade of the agent does not block the
installation of the auto-updater itself, so user might experience that the
agent is installed or upgraded **shortly after** the auto-updater being
installed.  *-- end note]*

Debian: ``sudo dpkg-query -W driveclient`` shows the installed version.

Redhat: ``sudo yum info driveclient`` shows the installed version.

If 3) is true, do nothing.

Otherwise, if 4) is not true, the installed agent is updated to the latest
version from the repository.
The actual update may be delayed for some time if the updater is running
as a daemon.  See `Usage`_ for details.

Use the command showing the installed version to compare the agent version
before and after the installation of the auto-updater.

*[Note:* To create a situation of "an old version of the agent is installed
without a configured agent repository", download an old deb or rpm agent
package and install them.  To create a situation of "an old version of the
agent is installed from the repository", you can downgrade the agent:

Debian: ``sudo apt-get install driveclient=<an old version>`` (Does not work unless you have the old version installed before.  Use the usual way to get this situation (i.e., install a local package).

Redhat: ``sudo yum downgrade driveclient``

*-- end note]*

Otherwise, if 5) is not true, then stop the agent daemon, upgrade the agent,
and start the new agent daemon.
The actual update may be delayed for some time if the updater is running
as a daemon.  See `Usage`_ for details.

ALL: If the logging level is set to `verbose` (already set in init script. See `Other Options`_ for details), the detailed steps will be logged (defaults to `/var/log/cloudbackup-updater.log`).  System log (`/var/log/messages`) may also contains the information about how the agent daemon is stopped and started, but the info text various on different system.

Otherwise, the auto-updater waits for the agent to finish the current backup,
restore, or cleanup task, stop the agent daemon, upgrade the agent, and start
the new agent daemon.

ALL: The agent holds a lock file, `/var/cache/driveclient/backup-running.lock`, when performing these tasks, so do the auto-updater, when performing **agent upgrade**.


Usage
=====

The auto-updater can be used in two ways:

1. As a command line tool.  Execute ``cloudbackup-updater`` from the command
   line without the **-d** option, then ``driveclient`` is installed or updated
   to the latest version.

2. As a daemon.  Execute ``cloudbackup-updater -d``, then ``driveclient`` is
   installed or updated to the latest version.  The auto-updater checks for the
   new agent version (as specified in `release/windows/version.txt` under the
   remote agent repository) every hour after it starts and keep the agent
   installed and updated.  The interval of the checking can be changed.  See
   `Other Options`_.

Only one ``cloudbackup-updater`` can run at the same time.  The attempt to
start a second daemon will fail with an error.

The command line ``cloudbackup-updater`` can be safely used when the daemon is
running.  One of them will perform the actual agent upgrade.  User can use the
command line to update the agent immediately without waiting for the next
checking.

When running as a daemon, after the auto-updater finds out that the latest
agent is newer than the installed one, the update may be delayed.  *50%* of
the users are expected to get the update within 6 hours, and *90%* of the
users are expected to get it within 11 hours.  If the environment variable
``CLOUDBACKUP_UPDATER_DEBUG`` is set, the unit is minute rather than hour.


Other Options
=============

**-r** `URI`: The remote agent repository, defaults to `http://agentrepo.drivesrvr.com`.  The local package management repositories are created according to this base URI.  Can be set to an internal address for testing.

**-v**: Turn on verbose logging.

**-h**: Display help messages.

The following options only affect the daemon mode.  They are simply ignored
in command line mode.

**-i** `NUM`: The new version checking interval in minutes, defaults to 60 (an hour).  If `NUM` is not an integer, fail with an error.

**-l** `PATH`: Path to the logging file.  The file will be created if not exist, but the directory part of `PATH` shall exists.

**-p** `PATH`: Path to the pidfile, the traditional UNIX daemon lock file.  The directory part of `PATH` shall exists.


debian/changelog
cloudbackup-updater (1.2-1) unstable; urgency=low

  * Update to 1.2

 -- Zhihao Yuan <zhihao.yuan@rackspace.com>  Mon, 08 Dec 2014 20:32:53 +0000

cloudbackup-updater (1.1-1) unstable; urgency=low

  * Update to 1.1

 -- Zhihao Yuan <zhihao.yuan@rackspace.com>  Thu, 02 Oct 2014 18:14:36 +0000

cloudbackup-updater (1.0-1) unstable; urgency=low

  * Update to 1.0

 -- Zhihao Yuan <zhihao.yuan@rackspace.com>  Wed, 04 Jun 2014 14:49:16 +0000

cloudbackup-updater (0.1-2) unstable; urgency=low

  * Switch to non-standard python packaging.

 -- Zhihao Yuan <zhihao.yuan@rackspace.com>  Thu, 30 Jan 2014 21:07:18 +0000

cloudbackup-updater (0.1-1) unstable; urgency=low

  * source package automatically created by stdeb 0.6.0+git

 -- Zhihao Yuan <zhihao.yuan@rackspace.com>  Fri, 10 Jan 2014 15:21:15 -0600


cloudbackup_updater/__init__.py
#!/usr/bin/env python

import os
import signal
import sys
import time
from getopt import getopt
import logging
import logging.handlers
import subprocess
import urllib2
import random
from string import Template

from ctxsoft import *
import dotlock
import pkgup
import pydaemon


VERSION_FILE_TMPL = (
    '%s/version.txt'
)

LOCK_FILE_TMPL = (
    '%s/backup-running.lock'
)

KEY_FILE_TMPL = (
    '%s/debian/agentrepo.key'
)

LOG = logging.getLogger()


def keep_upgraded(interval, url):
    while 1:
        if run_in_process(try_upgrade, url, return_to_wait=True):
            do_wait()
            run_in_process(try_upgrade, url)

        time.sleep(interval * 60)


def run_in_process(f, *args, **kwargs):
    try:
        ef = move_exitfunc()
        pid = os.fork()

        if pid == 0:
            st = 0
            try:
                try:
                    st = f(*args, **kwargs)

                except Exception, e:
                    LOG.exception(e)

            finally:
                sys.exit(st)

        else:
            set_exitfunc(ef)
            _, status = os.waitpid(pid, 0)

            if os.WIFEXITED(status):
                return os.WEXITSTATUS(status)

    except OSError, e:
        LOG.exception(e)


def move_exitfunc():
    try:
        f = sys.exitfunc
        del sys.exitfunc

        return f

    except AttributeError:
        return None


def set_exitfunc(f):
    if f is not None:
        sys.exitfunc = f


def try_upgrade(url, return_to_wait=False):
    vr_txt = remote_version(url).strip()

    try:
        vr = version_triple(vr_txt)

    except ValueError:
        raise RuntimeError("remote version.txt is malformed: %s" % vr_txt)

    except Exception, e:
        LOG.exception(e)

    if pkgup.PKG_MGMT == 'yum':
        get_repository = lambda: pkgup.Repository(name='drivesrvr')
        add_repository = add_yum_repository

    elif pkgup.PKG_MGMT == 'apt':
        get_repository = lambda: pkgup.Repository(name='serveragent')
        add_repository = lambda u: add_apt_repository('serveragent', u)

    try:
        repo = get_repository()

    except pkgup.NoSuchRepo:
        add_repository(url)
        repo = get_repository()

    pkg = repo.package('driveclient')
    cachedir = '/var/cache/driveclient'
    backup_lock = dotlock.DotLock(LOCK_FILE_TMPL % cachedir)

    try:
        vl_txt = pkg.installed_version()

    except pkgup.NotInstalled:
        LOG.info('Agent not installed')
        pkg.install()
        LOG.info('%s is newly installed', pkg.installed_nvra())

    else:
        if version_triple(vl_txt) < vr:
            if return_to_wait:
                return 1

            LOG.info('Agent version is behind: %s', vr_txt)

            def update():
                @with_(driveclient_not_running())
                def _as():
                    pkg.update(to=vr_txt)

            if os.path.isdir(cachedir):
                @with_(backup_lock)
                def _as():
                    LOG.info('Agent is idle; updating...')
                    update()

            else:
                update()

            LOG.info('%s is upgraded', pkg.installed_nvra())

    return 0


class driveclient_not_running(object):
    def __enter__(self):
        subprocess.call(['service', 'driveclient', 'stop'])
        LOG.info('Agent brought down')

    def __exit__(self, type, value, traceback):
        subprocess.call(['service', 'driveclient', 'start'])
        LOG.info('Agent brought up')


def add_yum_repository(url):
    uri = '%s/redhat/drivesrvr.repo' % url
    rfp = urllib2.urlopen(uri)

    if not 200 <= rfp.code < 300:
        raise RuntimeError('Failed to get repo file %r' % uri)

    @with_(closing(open('/etc/yum.repos.d/drivesrvr.repo', 'w')))
    def _as(fp):
        fp.write(rfp.read())
        LOG.info('Adding yum repository')


def add_apt_repository(name, url):
    add_apt_key(KEY_FILE_TMPL % url)

    @with_(closing(open('/etc/apt/sources.list.d/driveclient.list', 'w')))
    def _as(fp):
        fp.write('deb [arch=amd64] %s/debian/ %s main\n' % (url, name))
        LOG.info('Adding apt repository')


def add_apt_key(uri):
    fp = urllib2.urlopen(uri)

    if not 200 <= fp.code < 300:
        raise RuntimeError('Failed to get keyfile %r' % uri)

    p = subprocess.Popen(['apt-key', 'add', '-'],
                         stdin=subprocess.PIPE,
                         stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT)

    out = p.communicate(fp.read())[0]

    if p.returncode == 0:
        LOG.info('Adding apt key')

    else:
        LOG.error('Failed to add apt key:\n%s', out)
        sys.exit(1)


def remote_version(url):
    remote_version_file = VERSION_FILE_TMPL % url
    fp = urllib2.urlopen(remote_version_file)

    if not 200 <= fp.code < 300:
        raise RuntimeError('Failed to communicate %r' % remote_version_file)

    return fp.read()


def version_triple(version_text):
    return map(int, version_text.split('.'))


def do_wait():
    # E(x) = 6, Mode(x) = 4, at ~14%, Max(x) ~20
    # 10% ~ 2.2, 25% ~ 3.4, 50% ~ 5.3, 75% ~ 7.8, 90% ~ 10.6
    while True:
        x = random.gammavariate(3., 2.)
        if x < 24.:  # CDF(24) ~ 0.9995
            break
    if os.getenv('CLOUDBACKUP_UPDATER_DEBUG') is None:
        time.sleep(x * 3600)
    else:
        LOG.debug('Wait for %.1f minutes', x)
        time.sleep(x * 60)


def main_quit(signum=0, frame=None):
    if signum:
        LOG.info('Exit on signal %d' % signum)

    sys.exit()


def main_exec(cmd, *args):
    daemon_mode = False
    interval = 60
    cmd_name = os.path.basename(cmd)
    logfile = '/var/log/cloudbackup-updater.log'
    pidfile = '/var/run/cloudbackup-updater.pid'
    remote_prefix = 'http://agentrepo.drivesrvr.com'

    try:
        for k, v in getopt(args, 'di:l:p:r:vh')[0]:
            if k == '-h':
                print Template('''$cmd_name: [options]...
options:
  -d       daemon mode
  -i NUM   interval in minutes (defaults to $interval)
  -l PATH  path to log file (defaults to $logfile)
  -p PATH  path to pid file (defaults to $pidfile)
  -r URI   alternative remote repository
  -v       verbose logging
  -h       display this help''').substitute(locals())
                sys.exit(2)

            elif k == '-v':
                LOG.setLevel(logging.DEBUG)

            elif k == '-d':
                daemon_mode = True

            elif k == '-i':
                interval = int(v)

            elif k == '-l':
                logfile = v

            elif k == '-p':
                pidfile = v

            elif k == '-r':
                remote_prefix = v

    except Exception, e:
        print >> sys.stderr, '%s: %s' % (cmd_name, e)
        sys.exit(2)

    fmt = logging.Formatter('%(asctime)s [%(levelname)s]: %(message)s')

    if daemon_mode:
        d = pydaemon.Daemon(umask=077, pidfile=pidfile, verbose=0)

        def _as():
            log_handler = logging.handlers.RotatingFileHandler(
                logfile, maxBytes=4096, backupCount=5)
            log_handler.setFormatter(fmt)
            LOG.addHandler(log_handler)

            signal.signal(signal.SIGTERM, main_quit)

            keep_upgraded(interval, remote_prefix)

        d.run = _as
        d.start()

    else:
        log_handler = logging.StreamHandler()
        log_handler.setFormatter(fmt)
        LOG.addHandler(log_handler)

        try:
            try_upgrade(remote_prefix)

        except KeyboardInterrupt:
            pass


def main():
    main_exec(*sys.argv)
